### PR TITLE
Implement better rack awareness rebalance

### PR DESCRIPTION
**Pull request by antban**
ARUHA-1797

### CODE REVIEW

File: bubuku/features/rebalance/change_simple.py [hunk 1]
```diff
@@ -0,0 +1,443 @@
...
+        self._add_partition_int(partition.get_key(), leader)
+        return True
+
+
+def transfer_partition(partition, from_: Broker, target_brokers_sorted: List[Broker],
+                       zone_checker: ZoneChecker, polite=False):
+    for target in target_brokers_sorted:
+        if not zone_checker.is_allowed_move(from_, target, partition):
+            continue
+        if polite and target.is_overloaded(False, 1):
```
Comment (adyach) [inline bubuku/features/rebalance/change_simple.py]: should you actually pass the value from the caller function? it is not if it is leader or not at that point

File: bubuku/features/rebalance/change_simple.py [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+        leaders_load = sorted(self.active_brokers.values(), key=lambda x: x.get_load_factor(True))
+        replica_load = sorted(self.active_brokers.values(), key=lambda x: x.get_load_factor(False))
+
+        for pkey, is_leader in self.fake.list_partition_keys():
+            load = leaders_load if is_leader else replica_load
+            partition = self.partitions[pkey]
+            if not transfer_partition(partition, self.fake, load, self.zone_checker, polite=False):
+                raise Exception('Nowhere to transfer partition {}'.format(pkey))
+            self.register_partition_change(partition)
+            if is_leader:
```
Comment (adyach) [inline bubuku/features/rebalance/change_simple.py]: why do you need this `if` clause? you also pass `x.get_load_factor(True)` `True` in both cases

File: bubuku/features/rebalance/change_simple.py [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+                leaders_load = sorted(leaders_load, key=lambda x: x.get_load_factor(True))
+            else:
+                replica_load = sorted(replica_load, key=lambda x: x.get_load_factor(True))
+
+    def optimize_replicas(self):
+        is_leader = False
+        current_load = sorted(self.active_brokers.values(), key=lambda x: x.get_load_factor(is_leader))
+        moved = True
+        while moved:
+            to_empty = current_load[len(current_load) - 1]
```
Comment (adyach) [inline bubuku/features/rebalance/change_simple.py]: -1

File: bubuku/features/rebalance/change_simple.py [hunk 4]
```diff
@@ -0,0 +1,443 @@
...
+            if not to_empty.is_overloaded(is_leader):
+                return
```
Comment (adyach) [inline bubuku/features/rebalance/change_simple.py]: I thought you should check if optimisations of replicas possible for all underloaded brokers

File: bubuku/features/rebalance/change_simple.py [hunk 5]
```diff
@@ -0,0 +1,443 @@
...
+        new_broker_ids = sorted([str(id_) for id_ in self.zk.get_broker_ids()])
+        if new_broker_ids != self.initial_broker_ids:
+            _LOG.warning("Rebalance stopped because of broker list change from {} to {}".format(
+                self.initial_broker_ids, new_broker_ids))
+            return False
+
+        if self.state == SimpleRebalanceChange._STATE_INIT:
+            # Load data from zk
+            self.loadDataFromZk()
+            if not self.active_brokers:
```
[the remark this answers is not in the document]
Comment (antban) [reply bubuku/features/rebalance/change_simple.py]: There will be Exception "Nowhere to transfer partition" during empty_fake phase

**Comment by adyach** [comment]
👍

**Comment by antban** [comment]
:+1:

**Pull request merged**